helix-core/src/test/java/org/apache/helix/controller/rebalancer/TestAutoRebalanceStrategy.java

```diff
@@ -866,4 +907,220 @@ public void testWontMoveSinglePartitionUnnecessarily() {
+          "invalid preference list for " + partition);
+      Assert.assertEquals(znRecord.getMapField(partition).get(NODES[1]), STATE_MODEL.getTopState());
+      Assert.assertTrue(znRecord.getListField(partition).contains(NODES[0]),
+          "invalid preference list for " + partition);
+      Assert.assertEquals(znRecord.getMapField(partition).get(NODES[0]),
+          STATE_MODEL.getInitialState());
+    }
+  }
+
+  @Test
+  public void testRebalanceWithErrorPartition() {
+    final String RESOURCE_NAME = "resource";
+    final String[] PARTITIONS = {"resource_0", "resource_1", "resource_2"};
+    final StateModelDefinition STATE_MODEL = LeaderStandbySMD.build();
+    final int REPLICA_COUNT = 2;
+    final String[] NODES = {"n0", "n1", "n2", "n3", "n4", "n5", "n6", "n7", "n8", "n9"};
+
+    ResourceControllerDataProvider dataCache = buildMockDataCache(RESOURCE_NAME,
+        ResourceConfig.ResourceConfigConstants.ANY_LIVEINSTANCE.toString(), "LeaderStandby",
+        STATE_MODEL, Collections.emptySet());
+    // initial state, 10 node, no mapping
+    List<String> allNodes = Lists.newArrayList(NODES);
+    List<String> liveNodes = Lists.newArrayList(NODES);
+    Map<String, Map<String, String>> currentMapping = Maps.newHashMap();
+    for (String partition : PARTITIONS) {
+      currentMapping.put(partition, new HashMap<String, String>());
+    }
+
+    // make sure that when nodes join, all partitions is assigned fairly
+    List<String> partitions = ImmutableList.copyOf(PARTITIONS);
+    LinkedHashMap<String, Integer> stateCount =
+        STATE_MODEL.getStateCountMap(liveNodes.size(), allNodes.size());
+    ZNRecord znRecord =
+        new AutoRebalanceStrategy(RESOURCE_NAME, partitions, stateCount).computePartitionAssignment(
+            allNodes, liveNodes, currentMapping, dataCache);
+    Map<String, List<String>> preferenceLists = znRecord.getListFields();
+    for (String partition : currentMapping.keySet()) {
+      List<String> preferenceList = preferenceLists.get(partition);
+      Assert.assertNotNull(preferenceList, "invalid preference list for " + partition);
+      Assert.assertEquals(preferenceList.size(), allNodes.size(),
+          "invalid preference list for " + partition);
+    }
+
+    // Suppose that one replica of partition 0 is in n0, and it has been in the ERROR state.
+
+    for (String partition : PARTITIONS) {
+      Map<String, String> idealStateMap = znRecord.getMapField(partition);
+      currentMapping.put(partition, idealStateMap);
+    }
+    currentMapping.get(PARTITIONS[0]).put(NODES[0], "ERROR");
+    // now disable node 0, and make sure the dataCache provides it. And add another node n10 to the
+    // cluster. We want to make sure the n10 can pick up another replica of partition 0,1,2.
+    allNodes = new ArrayList<>(allNodes);
+    liveNodes = new ArrayList<>(liveNodes);
+    liveNodes.remove(NODES[0]);
+    allNodes.add("n10");
```

Comment:
Before we add another node, can we also check that the we don't drop a partition on any of the remaining nodes after disabling the one with the error replica?

Reply:
TFTR, added that check.